pennylane/workflow/qnode.py

```diff
@@ -96,13 +99,92 @@ def _make_execution_config(
 
     return qml.devices.ExecutionConfig(
         interface=getattr(circuit, "interface", None),
-        gradient_method=_gradient_method,
+        gradient_keyword_arguments=gradient_kwargs,
+        gradient_method=diff_method,
         grad_on_execution=grad_on_execution,
         use_device_jacobian_product=execute_kwargs.get("device_vjp", False),
         mcm_config=mcm_config or qml.devices.MCMConfig(),
     )
 
 
+def _update_mcm_config(mcm_config: "qml.devices.MCMConfig", interface: str, finite_shots: bool):
+    """Helper function to update the mid-circuit measurements configuration based on
+    execution parameters"""
+    if interface == "jax-jit" and mcm_config.mcm_method == "deferred":
+        # This is a current limitation of defer_measurements. "hw-like" behaviour is
+        # not yet accessible.
+        if mcm_config.postselect_mode == "hw-like":
+            raise ValueError(
+                "Using postselect_mode='hw-like' is not supported with jax-jit when using "
+                "mcm_method='deferred'."
+            )
+        mcm_config.postselect_mode = "fill-shots"
+
+    if (
+        finite_shots
+        and "jax" in interface
+        and mcm_config.mcm_method in (None, "one-shot")
+        and mcm_config.postselect_mode in (None, "hw-like")
+    ):
+        mcm_config.postselect_mode = "pad-invalid-samples"
+
+
+def _resolve_execution_config(
+    tapes: QuantumScriptBatch,
+    device: "qml.devices.Device",
+    execution_config: "qml.devices.ExecutionConfig",
+    transform_program: TransformProgram,
+) -> "qml.devices.ExecutionConfig":
+    """Resolves the execution configuration for non-device specific properties.
+
+    Args:
+        tapes (QuantumScriptBatch): a batch of tapes
+        device (qml.devices.Device): a Pennylane device
+        execution_config (qml.devices.ExecutionConfig): an execution config to be executed on the device
+        transform_program (TransformProgram): a program of transformations to be applied to the tapes
+
+    Returns:
+        qml.devices.ExecutionConfig: resolved execution configuration
+    """
+    if (
+        device.name == "lightning.qubit"
+        and qml.metric_tensor in transform_program
+        and execution_config.gradient_method == "best"
+    ):
+        gradient_fn = qml.gradients.param_shift
+    else:
+        gradient_fn = QNode.get_gradient_fn(

```

Comment:
Potentially worth relying on something else so that we can deprecate `get_gradient_fn`.

Reply:
Yeah ... I tried to use the new `qml.workflow._get_gradient_fn` earlier but it caused some tests to fail. I'm taking a look at it now and trying to make it work. 😄 

The problem is with this branch in `_get_gradient_fn`,
```python
...
    if diff_method == "best":
        qn = qml.QNode(lambda: None, device, diff_method=None)
        return qml.workflow.get_best_diff_method(qn)()
...
```
Since `get_best_diff_method` has no direct access to the tape provided to `_get_gradient_fn` it incorrectly returns `backprop` when it should be `parameter-shift` for finite shots. I tried setting the tape with `qn._tape = tape` but that didn't work. 

Any ideas?

Reply:
I think the best way forward would be to modify the `qml.workflow._get_gradient_fn` to not rely on `qml.workflow.get_best_diff_method` and instead replace that block of code with the necessary logic from the latter,
```python
...
    if diff_method == "best":
        config = _make_execution_config(None, "best")

        if device.supports_derivatives(config, circuit=tape):
            new_config = device.preprocess(config)[1]
            return new_config.gradient_method

        if tape and any(isinstance(o, qml.operation.CV) for o in tape):
            return qml.gradients.param_shift_cv
            
        return qml.gradients.param_shift
 ...
```

Reply:
I think you just need to do something like pass shots from the tape in when calling the output of `get_best_diff_method` inside `_get_gradient_fn` and that should resolve. 

I.e. something along the lines of `return qml.workflow.get_best_diff_method(qn)(shots=tape.shots)`

Reply:
Interesting, I can try that.

Although, this might actually be an improvement to `_get_gradient_fn` since `get_best_diff_method` is designed to be user-friendly - returning strings rather than transform functions. Seems kind of weird for an internal developer function to rely on a user-designed function? 

By extracting the relevant code from `get_best_diff_method`, we could simply return the transform function directly.

Reply:
I can see that it might be nice if `_get_gradient_fn` returned a function instead of a string. I don't have a strong opinion either way - since we are allowing users to pass in `diff_method`, we are going to have a mix of strings and functions to define our gradients up to some point, and I _think_ at least in our current implementation we still have a mix at this point.